conan_provider.cmake

```diff
@@ -275,7 +275,8 @@ macro(conan_provide_dependency package_name)
     if(${index} EQUAL -1)
         list(PREPEND CMAKE_PREFIX_PATH "${CONAN_GENERATORS_FOLDER}")
     endif()
-    find_package(${ARGN} BYPASS_PROVIDER CMAKE_FIND_ROOT_PATH_BOTH)
+    set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE BOTH)
```

Comment:
The provider captures every ``find_package()``, so I don't see how restoring the value after the call would make any difference, so I think it is good as-is.

Reply:
I think it's probably best to leave `CMAKE_FIND_ROOT_PATH_MODE_PACKAGE` untouched, regardless of whether we set it in the call to `find_package`, or before as in this PR.

When cross building, the `BOTH` value gives priority and re-roots the entire search order around `CMAKE_FIND_ROOT_PATH` and `CMAKE_SYSROOT` if defined.  So if the Sysroot contains CMake configuration files, it will look for those FIRST, before looking in the `CONAN_GENERATORS_FOLDER` which is what we want. This is likely to happen on sysroots for Linux distros, like Raspberry Pi or NVIDIA Jetson.


The issue with the Android NDK which motivated this workaround, as reported in this previous PR: https://github.com/conan-io/cmake-conan/pull/521#discussion_r1227625701 - has to do with the fact that the Android NDK toolchain sets this to `ONLY`, which causes find_package to never consider anything outside of the sysroot: https://github.com/android/ndk/issues/517

When it comes to the specific variable `CMAKE_FIND_ROOT_PATH_MODE_PACKAGE`, this is probably a defect/bug or oversight in the Android NDK toolchain: there aren't any packages in the sysroot - so while this approach may be valid for libraries and headers, it is not for CMake packages, unless the expectation is that users add files to the Android sysroot:

```
/opt/android_ndk_r25b/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr
luisc@luisc-mac usr % find . -name "*.cmake"
```
^ (the following returns nothing)


If we set this to `BOTH` - when crossbuilding we would be in essence recreating the problems described in this issue: https://github.com/conan-io/conan/issues/12478 - giving more priority to the sysroot than to Conan.

We also cannot set this to `NEVER`, because it's not entirely uncommon to expect some packages to be provided by the SDK/sysroot.

So taking into consideration that we do indeed capture all calls to `find_package`, we probably need a logic that ensures Conan is searched first, and if the package is not found, then it follows normal procedure. CMake's [documentation](https://cmake.org/cmake/help/latest/command/find_package.html) does suggest an option:
```
find_package (<PackageName> PATHS paths... NO_DEFAULT_PATH)
find_package (<PackageName>)
```

which in our case could be translated to:
```
find_package (<PackageName> ${ARGN} PATHS "${CONAN_GENERATORS_FOLDER}" NO_DEFAULT_PATH)
find_package (<PackageName> ${ARGN})
```

The first call, in essence, prevents CMake from looking anywhere BUT `CONAN_GENERATORS_FOLDER` - and only if it is not found there, would the second call come into play. 

We have a secondary issue, which Im not sure it's been reported yet, but our dependency provider will fail for any dependency where the `package_info()` in the recipe is set to  *ONLY* generate a find_module.

Reply:
This isn't just an Android toolchain issue, iOS/tvOS/watchOS has the same issue. Before posting this fix, I undid the `CMAKE_FIND_ROOT_PATH_BOTH` setting on https://github.com/conan-io/cmake-conan/pull/528 and got a similar error as we did on Android:
https://github.com/conan-io/cmake-conan/actions/runs/5581115133/jobs/10198913086

From here, I think we just have try solutions against the existing test coverage like your suggestion above, keep adding any missing test coverage, and see what solution satisfies them all. I'm happy to help with reviewing and testing.

I do think this merge did get us closer to working for all scenarios so, in my opinion it should stay on `develop2` while further improvements are made.

Reply:
Thanks for the additional info!

I said add test coverage not just use existing coverage and I certainly agree on proactively solving problems. I'm even willing to work on it though I'd prefer to finish up the two other pending PRs first, if possible.

Reply:
Oh, thank you for discovering and bringing attention to these issues in the first place! It's a great motivator to move this tool forward :D 

The other two pending PRs are on my priority list this week :)